CI: improve test suite runtime via pytest parallelism and disabling mypy in most jobs

rgommers opened the pull request:
Should address some of the issues identified in [gh-24280](https://github.com/numpy/numpy/issues/24280#issuecomment-1657094780).

Regarding the Mypy tests: these tests are super slow, and they're effectively always passing in CI. Running them on all "full" test suite runs is too expensive, so enable them on a single job instead. Note that SciPy has an XSLOW mark, NumPy does not. So use an env var for now. Typical guidelines are that fast tests should be >0.5 sec, slow only in the 0.5-5 sec range, and anything above that should be avoided. The boundaries are fuzzy, but Mypy takes 1-2 minutes to run, so needs custom opt-in. Cc @BvB93 for visibility.

Also adds a `spin mypy` developer command to make it easier to run only static typing tests.

rgommers committed:
CI: ensure tests in cibuildwheel jobs run in parallel

[skip azp] [skip circle] [skip travis]

rgommers committed:
TST: disable mypy tests in test suite unless an environment variable is set

These tests are super slow, and they're effectively always passing in CI.
Running them on all "full" test suite runs is too expensive. Note that
SciPy has an XSLOW mark, NumPy does not. So use an env var for now.

[skip circle] [skip travis] [skip azp]

BvB93 commented:
Considering we have a dedicated static typing label, would it be an idea to automatically enable to env var if it's set? Situations wherein the label is present but one is _not_ interested in running the mypy are, realistically speaking, non existent.

rgommers commented:
> Considering we have a dedicated static typing label, would it be an idea to automatically enable to env var if it's set? 

That seems like a good idea to me. In case that is hard to do on, e.g., Azure (since I don't think we have any label-based logic there) we can perhaps also recognize `[run mypy]` in the commit message.

rgommers commented:
The Intel SPR failures seem unrelated:
```
AssertionError: FPU precision mode changed from 0x77f to 0x37f during the test
```

The macOS arm64 wheel build jobs looks much better:

<img width="597" alt="image" src="https://github.com/numpy/numpy/assets/98330/b9cc9e7f-a9db-47d2-b94c-89dbcce47157">

So this almost halved the runtime compared to before (see [here](https://github.com/numpy/numpy/issues/24280#issuecomment-1657094780)). 

The _Run cibuildwheel_ step duration decreased by more than 2x, and average CPU utilization increased from ~1 to ~2.5 (out of 4 cores) during that whole phase.

rgommers commented:
And same for the Linux aarch64 jobs, runtime about halved, from 19 min to 9 min 30s on average:

<img width="597" alt="image" src="https://github.com/numpy/numpy/assets/98330/b89449ae-b656-4841-b42a-f5caf2feddf9">

BvB93 commented:
> ...Azure (since I don't think we have any label-based logic there) we can perhaps also recognize `[run mypy]` in the commit message.

Hm, a shame that it cannot be done automatically for Azure, but so be it; directly using the commit message is a decent plan B.

Just brain storming here at this point, but how feasible would it be to automatically run it if a change is detected in one of the .pyi files?

rgommers commented:
> Just brain storming here at this point, but how feasible would it be to automatically run it if a change is detected in one of the .pyi files?

On GitHub Actions one can provide a list of paths to trigger on if those are touched in the PR diff. It'd be a pretty long list, but it's doable. I think on Azure it's again more tricky. 

Let me ask a question in return though: what kind of coverage actually makes sense for these tests? I imagine that we'd want macOS/Linux/Windows and also all covered Python versions (currently 3.9-3.12). Does it have to be a dense matrix though across all CI providers? It's hard to see why that would matter. So maybe what would be easier is to ensure we cover that set of OSes and Python versions on GitHub Actions? 

Or maybe we can use `pyright` instead? It's several times faster typically.

BvB93 commented:
> Let me ask a question in return though: what kind of coverage actually makes sense for these tests?

You make a good point here: all that's really important is to check the basic mac/windows/unix + Python version combinations. But anything else? Tests with specific compiler options, 32 bit windows, Etc.? They're all, at minimum, fairly irrelevant here for the purpose of typing related tests and disabling them in full on those platforms is perfectly fine IMO.

As for switching to Pyright? This will be a non-trivial amount of work as our test suite is currently heavily specialised for mypy. There is https://github.com/numpy/numpy/issues/21505 on the wish list that should make adapting the test suit for different type checkers a bit easier, but again: that's just on the wish list for now.

rgommers commented:
> You make a good point here: all that's really important is to check the basic mac/windows/unix + Python version combinations

Thanks for confirming. I'm considering putting that in a separate GHA yaml file - then it's also triggerable by hand via "workflow dispatch" for any maintainer, and we can keep all the paths to `.pyi` files in a single list. And in addition we have single mypy run in regular CI for PRs that don't touch `.pyi` files. Running mypy takes about as much time as a full build, so it's not like the new/extra jobs would add a ton of overhead.

BvB93 commented:
GH also supports wild cars containing file paths for its trigger list, so it's probably easier to just use `**.pyi` over explicitly enumerating them.

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-paths

rgommers commented:
I'm marking this as draft until the new Mypy-specific CI jobs have been added. I am backporting the changes here to the `1.26.x` branch, because there we don't need those new jobs, the only purpose of that branch is build system changes and a new release, so cutting the wheel build time by ~2x is very useful there.

rgommers committed:
CI: add new GHA CI jobs to run MyPy across OS/Python flavors

[skip cirrus] [skip travis] [skip circle] [skip azp]

rgommers committed:
DEV: add a `spin mypy` command

rgommers commented:
This should be good to go now:

- I added a `spin mypy` command for convenience, and
- added 3 dedicated jobs (runtime is only ~15 minutes combined, so should be okay) to run the `numpy/typing` tests on Linux, macOS and Windows with 3 different Python versions. 
- Those jobs run by default, and only for PRs that touch some clearly unrelated folders like `docs/` the runs will be skipped.

mattip commented on tools/wheels/cibw_test_command.sh:
```diff
@@ -26,5 +26,7 @@ fi
-python -c "import sys; import numpy; sys.exit(not numpy.test(label='full'))"
+# Run full tests with -n=auto. This makes pytest-xdist distribute tests across
+# the available N CPU cores: 2 by default for Linux instances and 4 for macOS arm64
+python -c "import sys; import numpy; sys.exit(not numpy.test(label='full', extra_argv=['-n=auto']))"
```
Are you sure it is worthwhile distributing tests when there are only 2 cores? Does the total time go down?

rgommers replied:
yes, it's faster. typically ~50% or so

mattip commented:
[This comment](https://github.com/numpy/numpy/pull/24291#issuecomment-1657130514) mentions triggering on the "Static Typing" label and on a `[run mypy]` commit message. I don't see that in the mypy.yaml, is that follow-on work or is it too hard to do?

rgommers commented:
> [This comment](https://github.com/numpy/numpy/pull/24291#issuecomment-1657130514) mentions triggering on the "Static Typing" label and on a `[run mypy]` commit message. I don't see that in the mypy.yaml, is that follow-on work or is it too hard to do?

It's run by default already, so that's not needed. We shouldn't turn it off completely for regular Python/C work in the package, because that can break static typing

mattip commented:
Thanks @rgommers
